Mathlib/NumberTheory/SmoothNumbers.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2023 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.LinearAlgebra.ExteriorAlgebra.Grading
+import Mathlib.Data.Nat.Factorization.Basic
+
+/-!
+# Smooth numbers
+
+We define the set `Nat.smoothNumbers n` consisting of the positive natural numbers all of
+whose prime factors are strictly less than `n`.
+
+We also define the finite set `Nat.primesBelow n` to be the set of prime numbers less than `n`.
+
+The main definition `Nat.equivProdNatSmoothNumbers` establishes the bijection between
+`ℕ × (smoothNumbers p)` and `smoothNumbers (p+1)` given by sending `(e, n)` to `p^e * n`.
+Here `p` is a prime number.
+-/
+
+namespace Nat
+
+/-- `primesBelow n` is the set of primes less than `n` as a finset. -/
+def primesBelow (n : ℕ) : Finset ℕ := (Finset.range n).filter (fun p ↦ p.Prime)
+
+@[simp]
+lemma primesBelow_zero : primesBelow 0 = ∅ := rfl
+
+lemma prime_of_mem_primesBelow {p n : ℕ} (h : p ∈ n.primesBelow) : p.Prime :=
+  (Finset.mem_filter.mp h).2
+
+lemma lt_of_mem_primesBelow {p n : ℕ} (h : p ∈ n.primesBelow) : p < n :=
+  List.mem_range.mp <| Finset.mem_of_mem_filter p h
+
+lemma primesBelow_succ (n : ℕ) :
+    primesBelow n.succ = if n.Prime then insert n (primesBelow n) else primesBelow n := by
+  rw [primesBelow, primesBelow, Finset.range_succ, Finset.filter_insert]
+
+lemma not_mem_primesBelow (n : ℕ) : n ∉ primesBelow n := by
+  intro hn
+  simp [primesBelow] at hn
+
+/-- `smoothNumbers n` is the set of *`n`-smooth positive natural numbers*, i.e., the
+positive natural numbers all of whose prime factors are less than `n`. -/
+def smoothNumbers (n : ℕ) : Set ℕ := {m | m ≠ 0 ∧ ∀ p ∈ factors m, p < n}
+
+@[simp]
+lemma smoothNumbers_zero : smoothNumbers 0 = {1} := by
+  have h (m : ℕ) : (∀ p ∈ factors m, p < 0) ↔ m.factors = []
+  · rw [List.eq_nil_iff_forall_not_mem]
+    exact ⟨fun H p hp ↦ not_succ_le_zero p (H p hp), fun H p hp ↦ False.elim <| H p hp⟩
+  ext m
+  rw [smoothNumbers, Set.mem_setOf, h, factors_eq_nil,
+      (show m ∈ ({1} : Set ℕ) ↔ m = 1 from Set.mem_def), Ne.def]
+  exact ⟨fun ⟨H₁, H₂⟩ ↦ H₂.resolve_left H₁, fun H ↦ ⟨H.symm ▸ one_ne_zero, Or.inr H⟩⟩
```

Comment:
Alternative proof:
```suggestion
  ext m
  rw [Set.mem_singleton_iff, smoothNumbers, Set.mem_setOf]
  simp_rw [Nat.not_lt_zero]
  rw [← List.eq_nil_iff_forall_not_mem, factors_eq_nil, and_or_left, not_and_self_iff, false_or,
    ne_and_eq_iff_right zero_ne_one]
```
(Optional, but note in particular `Set.mem_singleton_iff`)

Reply:
I'll take the second one.